lua/starfall/libs_sh/quaternion.lua

```diff
@@ -1,681 +1,746 @@
 
---[[
--- Quaternion Support
--- Converted from Wiremod's E2 Quaternion library for general lua use
--- Original code for use by Bubbus
--- Permission received for use from Bubbus by Radon
--- http:\\wiki.wiremod.com/?title=Expression2#Quaternion
---
--- Credits to Radon for addition to Starfall
--- Credits to Divran for painful amounts of testing
-]]
-
--- faster access to some math library functions
-local math = math -- Because global lookups suck
-local setmetatable = setmetatable
-local abs   = math.abs
-local Round = math.Round
-local sqrt  = math.sqrt
-local exp   = math.exp
-local log   = math.log
-local sin   = math.sin
-local cos   = math.cos
-local sinh  = math.sinh
-local cosh  = math.cosh
-local acos  = math.acos
-local min 	= math.min
-
-local delta = 0.0000001000000
-
-local deg2rad = math.pi / 180
-local rad2deg = 180 / math.pi
-
-
---- Quaternion library
--- @name quaternion
--- @class library
--- @libtbl quat_lib
-SF.RegisterLibrary("quaternion")
-
---- Quaternion type
+--- Quaternion type. Recently reworked library, for full changelist visit: https://github.com/thegrb93/StarfallEx/pull/953
 -- @name Quaternion
 -- @class type
 -- @libtbl quat_methods
 -- @libtbl quat_meta
-SF.RegisterType("Quaternion")
+SF.RegisterType("Quaternion", true, false, nil, nil, function(checktype, quat_meta)
+	return function(q)
+		return setmetatable({ q:Unpack() }, quat_meta)
+	end,
+	function(obj)
+		checktype(obj, quat_meta, 2)
+		return {obj[1], obj[2], obj[3], obj[4]}
+	end	
+end)
 
 
 return function(instance)
 
-local quat_lib = instance.Libraries.quaternion
-local quat_methods, quat_meta = instance.Types.Quaternion.Methods, instance.Types.Quaternion
-local ent_meta, ewrap, eunwrap = instance.Types.Entity, instance.Types.Entity.Wrap, instance.Types.Entity.Unwrap
-local ang_meta, awrap, aunwrap = instance.Types.Angle, instance.Types.Angle.Wrap, instance.Types.Angle.Unwrap
-local vec_meta, vwrap, vunwrap = instance.Types.Vector, instance.Types.Vector.Wrap, instance.Types.Vector.Unwrap
+local checktype = instance.CheckType
+local quat_methods, quat_meta, qwrap, unwrap = instance.Types.Quaternion.Methods, instance.Types.Quaternion, instance.Types.Quaternion.Wrap, instance.Types.Quaternion.Unwrap
+local ent_methods = instance.Types.Entity.Methods
+local ang_methods, awrap, aunwrap = instance.Types.Angle.Methods, instance.Types.Angle.Wrap, instance.Types.Angle.Unwrap
+local vec_methods, vec_meta, vwrap, vunwrap = instance.Types.Vector.Methods, instance.Types.Vector, instance.Types.Vector.Wrap, instance.Types.Vector.Unwrap
+local mwrap = instance.Types.VMatrix.Wrap
+local math_library = instance.Libraries.math
+
+local function wrap(tbl)
+	return setmetatable(tbl, quat_meta)
+end
 
 local getent
 instance:AddHook("initialize", function()
 	getent = instance.Types.Entity.GetEntity
 end)
 
+-------------------------------------
 
---****************************** Helper functions ******************************--
+local math_sqrt = math.sqrt
```

Comment:
I'd move these locals and the below local functions up outside of the `return function()`